bindings/python/htcondor2/_credd.py

```diff
@@ -1 +1,208 @@
+from typing import List
 
+from ._common_imports import (
+    classad,
+    Collector,
+    DaemonType,
+    CredType,
+    CredCheck,
+)
+
+from .htcondor2_impl import (
+    _credd_do_store_cred,
+)
+
+
+class Credd():
+
+    def __init__(self, location : classad.ClassAd = None):
+        self._default = False
+
+        if location is None:
+            self._default = True
+
+            c = Collector()
+            # The version 1 documentation says that the we assume the
+            # local schedd, but the implementation uses the local credd.
+            location = c.locate(DaemonType.Credd)
+
+        if not isinstance(location, classad.ClassAd):
+            raise TypError("location must be a ClassAd")
+
+        self._addr = location['MyAddress']
+        # We never actually use this for anything.
+        # self._version = location['CondorVersion']
+
+
+    def add_password(self, password : str, user : str = None) -> None:
+        if len(password) == 0:
+            # This was HTCondorValueError in version 1.
+            raise ValueError("password may not be empty")
+
+        mode = self._STORE_CRED_LEGACY_PWD | self._GENERIC_ADD
+
+        addr = self._addr
+        if self._default:
+            addr = None
+
+        _credd_do_store_cred(addr, user, password, mode, None, None)
+
+
+    def delete_password(self, user : str = None) -> bool:
+        mode = self._STORE_CRED_LEGACY_PWD | self._GENERIC_DELETE
+
+        addr = self._addr
+        if self._default:
+            addr = None
+
+        result = _credd_do_store_cred(addr, user, None, mode, None, None)
+        return result == self._SUCCESS
+
+
+    def query_password(self, user : str = None) -> bool:
+        # FIXME: Test this.  A return of FAILURE_NOT_FOUND did
+        # not raise an exception in version 1.
+        mode = self._STORE_CRED_LEGACY_PWD | self._GENERIC_QUERY
+
+        addr = self._addr
+        if self._default:
+            addr = None
+
+        result = _credd_do_store_cred(addr, user, None, mode, None, None)
+        return result == self._SUCCESS
+
+
+    def add_user_cred(self, credtype : CredType, credential : bytes, user : str = None ) -> None:
+        mode = self._GENERIC_ADD
+        if credtype == CredType.Password:
+            mode |= credtype
+        elif credtype == CredType.Kerberos:
+            mode |= credtype | self._STORE_CRED_WAIT_FOR_CREDMON
+        else:
+            # This was HTCondorEnumError in version 1.
+            raise RuntimeError("invalid credtype")
+
+        if credential is None:
+            credential = _credd_run_credential_producer(
```

Comment:
I don't see _credd_run_credential_producer() defined anywhere.

Reply:
Pushing this off (with a demo implementation already written) into HTCONDOR-2163.